src/org/mirah/typer/typer.mirah

```diff
@@ -574,6 +569,7 @@ class Typer < SimpleNodeVisitor
     end
     enclosing_node = node.findAncestor {|n| n.kind_of?(MethodDefinition) || n.kind_of?(Script)}
     if enclosing_node.kind_of? MethodDefinition
+      return nil if isMethodInBlock(MethodDefinition(enclosing_node)) # return types are not supported currently for methods which act as templates
```

Comment:
I think it'll only cause endless recursion for the new code. In the existing version, the method's future is set at https://github.com/mirah/mirah/blob/master/src/org/mirah/typer/typer.mirah#L1153

We could make that more explicit by using `getInferredType` here, and `learnType` there, since neither will cause additional traversal.

Is it necessary to not give nested methods a future, or is it just that they don't need the future yet? Maybe we could still set a MethodFuture for them, but let it get overridden during the re-infer. Ultimately, I think we want to have all the type futures generated at this stage, and just copy them over for the closure building, but we can do that incrementally.